Temp Fix to restrict access for unconfirmed users

F3PiX opened the pull request:
Related issue #991 (Break Down step 2)

We need to restrict access for unconfirmed users but let them update their email address nevertheless. 
- Changed Ability class to catch unconfirmed users.
- Added a 'confirmed user' context for controller tests
- Updated the controller tests that failed; left others alone
- Extra: added first feature test, for guest user. 
Note: I would appreciate help with the feature tests.  This is a first attempt.
 
<!----Describe what this PR is about:
 - What feature does it add, which bug does it fix? 
 - Tests are much appreciated. 
 - Add screenshots if your PR includes visual/UI changes
---->
Follow up issues:
- [ ] check if we need the `can :read, :feed_entry`, or else remove its `can?` check in views. (Public activities shouldn't be restricted at all.)
- [ ] check controller tests and use either 'confirmed user' or 'unconfirmed user' contexts
- [ ] Fix pending specs and todo's in ability_spec

F3PiX commented on spec/models/ability_spec.rb:
```diff
@@ -69,6 +71,7 @@
               allow(user).to receive(:confirmed?).and_return(false)
             end
             it 'allows to see not hidden email address' do
+              pending "Fails. Unconfirmed user has no access"
```
??? This spec is failing, but I am not sure if we even want it to pass for this conditions?

ramonh replied:
Hmm. You're right. All an unconfirmed user should, in theory, be able to do is edit their email address.

F3PiX replied:
How about we keep it pending for now? I expect that it will solve itself in the upcoming Overhaul steps.

ramonh replied:
Yes, I agree. Is there a way to mark this as a TODO? Or is the `pending` enough, do you reckon?

F3PiX replied:
I am currently collecting all the todo's, and will add them when we have finished the discussion in #991 . Don't want to add to much detail to that.

ramonh replied:
Fair point!

F3PiX commented:
@pgaspar @klappradla @ramonh Can you take a look please?

ramonh commented on app/models/ability.rb:
```diff
@@ -10,102 +8,110 @@ def initialize(user)
-    can :crud, User, id: user.id
-    can :crud, User if user.admin?
+    can :read, User
+    can :update, User, id: user.id
```
I think the failing spec comes from here. Is there a reason we're removing admin access from `crud`'ing users?

F3PiX replied:
Admin can still crud User, it has moved to line 38.

ramonh replied:
My apologies! Dunno how I missed that...

carpodaster replied:
But the admin call is guarded behind a `if user.confirmed?`. I feel we should keep the current behaviour. If anything, we want admins to be able to fix a user's typo'd email address as a response to a support request ticket.

pgaspar commented on app/models/ability.rb:
```diff
@@ -10,102 +8,110 @@ def initialize(user)
-  def on_team_for_season?(user, season)
-    season && user.roles.student.joins(:team).pluck(:season_id).include?(season.id)
-  end
+      can :crud, :comments if user.admin?
```
This is not related to this PR specifically, and we should probably open a separate issue for this: it seems we're not testing this ability in the project comments (or anywhere I could find).

We used to have this in the teams show page, but it was removed meanwhile: https://github.com/rails-girls-summer-of-code/rgsoc-teams/blob/4fc8dfdec20a377fd3a1a500e2399aac7cea48d5/app/views/teams/show.html.slim#L17-L22

I've not used CanCanCan a lot in the past.. do we usually use the [controller helpers](https://github.com/CanCanCommunity/cancancan#3-controller-helpers) or just view helpers?

Should I create the issue or do we want to fix it in this PR?

F3PiX replied:
We use both helpers ( PR #993 is all about the controller helpers :-) )
If it's okay with you, I'd prefer to keep the PR's as clean as possible. And collect issues for every step layed out in the Overhaul: Breakdown (#991) . So we can use that as a reference all the way.
The `crud comments issue` will be part of the CommentsController Overhaul (listed in nr 5 in 991).  Or afterwards, in the restructuring of the Ability. I'll add it. Thanks!!

pgaspar replied:
Of course! Thank you for your patience - I'm still figuring out how we structure things! 😊

pgaspar commented:
I like these changes 👍 I didn't do a lot of testing, but I think the changes make sense so far!

I'm not sure how thoroughly we use these abilities though, there seem to be some that are not checked anywhere - which means unconfirmed users can do _some_ things (I didn't test much other than the comments, but that's one that users can use even before confirmation).

Not sure if we should make an effort to see if thee abilities are used or not (in a separate issue, I assume) or if it's too much trouble.

F3PiX commented:
I don't have authority in CodeClimate to set the warning to 'wontfix'. Can someone disable the warning please? (@ramonh ?)

ramonh commented:
Oh! Sure! I thought it might've been a comment. I'll take a look, as this PR otherwise looks solid to me. Thanks @F3PiX

ramonh commented:
I think I did it 👍 Let's see!

carpodaster commented:
@F3PiX I marked the CC issue as wontfix

F3PiX commented:
@carpodaster THANKS! and sorry for the mess.
1) Moved the helper_methods back home. 🙈 
2) the `if user.confirmed?` checks if the admin user herself is confirmed. Which I expected always to be true...?
3) `can :crud, User if user.admin?` should check if an admin can :crud each and every @user. Because we are not passing in a restriction with the user ids. 
(As opposed to a user who can only crud their own record:
`can :crud, User, id: user.id`)
4) I added very verbose expectations, to double check if we are testing what we want to test, and the behaviour we want. I think the user.admin crudding users is covered now?
5) All this reveals two bigger problems: I didn't update the specs properly, my mistake. And the specs are confusing. 
6) My proposal: close this PR for now, first make sure that we can trust the abilities_specs before changing the behaviour. 
WDYT?

F3PiX commented:
This is covered now by #997 
Admin specs are covered too:
https://github.com/rails-girls-summer-of-code/rgsoc-teams/blob/9d685da7aeed9045bf9d7de40e973bc782d2eb66/spec/models/ability/admin_spec.rb#L14-L24
Closing this.
